File: crates/breez-sdk/bindings/examples/cli/langs/react-native/src/advanced.ts
/**
 * Advanced subcommands.
 *
 * Mirrors the Rust CLI `advanced` subcommands:
 *   unilateral-exit, check-unilateral-exit, export-unilateral-exit-state,
 *   import-unilateral-exit-state
 */

import {
  singleKeyCpfpSigner,
  CpfpFundingKind,
  CpfpInput,
  ExitLeafSelection,
  ExitTransactionStatus_Tags,
  UnilateralExitVerdict_Tags,
} from '@breeztech/breez-sdk-spark-react-native'
import type {
  BreezSdkInterface,
} from '@breeztech/breez-sdk-spark-react-native'
import RNFS from 'react-native-fs'
import { formatValue } from './serialization'

function hexToArrayBuffer(hex: string): ArrayBuffer {
  const cleaned = hex.replace(/\s/g, '')
  const bytes = new Uint8Array(cleaned.length / 2)
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(cleaned.substring(i * 2, i * 2 + 2), 16)
  }
  return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength)
}

/** All advanced subcommand names for help and completion. */
export const ADVANCED_COMMAND_NAMES = [
  'unilateral-exit',
  'check-unilateral-exit',
  'export-unilateral-exit-state',
  'import-unilateral-exit-state',
]

/**
 * Parse a named flag value from args. Returns the value after the flag, or undefined.
 */
function parseFlag(args: string[], ...flags: string[]): string | undefined {
  for (const flag of flags) {
    const idx = args.indexOf(flag)
    if (idx !== -1 && idx + 1 < args.length) {
      return args[idx + 1]
    }
  }
  return undefined
}

/**
 * Parse a multi-value flag (comma-separated). Returns undefined if not provided.
 */
function parseMultiFlag(args: string[], ...flags: string[]): string[] | undefined {
  const val = parseFlag(args, ...flags)
  if (val === undefined) return undefined
  return val.split(',').map(s => s.trim()).filter(s => s.length > 0)
}

/**
 * Parse repeated flags (each occurrence has its own value).
 */
function parseRepeatedFlag(args: string[], ...flags: string[]): string[] {
  const values: string[] = []
  for (let i = 0; i < args.length; i++) {
    if (flags.includes(args[i]) && i + 1 < args.length) {
      values.push(args[i + 1])
      i++
    }
  }
  return values
}

function resolvePath(path: string): string {
  return path.startsWith('/')
    ? path
    : `${RNFS.DocumentDirectoryPath}/${path}`
}

async function readExitFile(path: string): Promise<any> {
  const json = await RNFS.readFile(path, 'utf8')
  return JSON.parse(json)
}

async function writeExitFile(path: string, exit: unknown): Promise<void> {
  const json = JSON.stringify(
    exit,
    (_key, value) => (typeof value === 'bigint' ? Number(value) : value),
    2
  )
  await RNFS.writeFile(path, json, 'utf8')
}

/**
 * Dispatch an advanced subcommand.
 *
 * @param args - The arguments after "advanced" (e.g., ["unilateral-exit", "--fee-rate", "2", ...])
 * @param sdk - The BreezSdkInterface instance
 * @returns A string result to display
 */
export async function dispatchAdvancedCommand(
  args: string[],
  sdk: BreezSdkInterface
): Promise<string> {
  if (args.length === 0 || args[0] === 'help') {
    return printAdvancedHelp()
  }

  const subcommand = args[0]
  const subArgs = args.slice(1)

  switch (subcommand) {
    case 'unilateral-exit':
      return handleUnilateralExit(sdk, subArgs)
    case 'check-unilateral-exit':
      return handleCheckUnilateralExit(sdk, subArgs)
    case 'export-unilateral-exit-state':
      return handleExportUnilateralExitState(sdk, subArgs)
    case 'import-unilateral-exit-state':
      return handleImportUnilateralExitState(sdk, subArgs)
    default:
      return `Unknown advanced subcommand: ${subcommand}. Use 'advanced help' for available commands.`
  }
}

function printAdvancedHelp(): string {
  const lines = [
    '',
    'Advanced subcommands (expert-only, misuse can strand or lose funds):',
    '  advanced unilateral-exit --fee-rate <rate> --destination <addr>',
    '    [--funding-kind p2wpkh|p2tr] [--leaf <id>,<id>,...]',
    '    [--utxo txid:vout:value:pubkey ...] [--secret-key <hex>]',
    '    [--output-file <path>]',
    '                                         Build and sign a unilateral exit',
    '  advanced check-unilateral-exit --input-file <path>',
    '    [--output-file <path>]',
    '                                         Check a signed exit against the chain',
    '  advanced export-unilateral-exit-state --output-file <path>',
    '                                         Export exit state to a file',
    '  advanced import-unilateral-exit-state --input-file <path>',
    '                                         Import exit state from a file',
    '',
  ]
  return lines.join('\n')
}

// --- unilateral-exit ---

function parseCpfpInput(s: string, kind: string): InstanceType<typeof CpfpInput.P2wpkh> | InstanceType<typeof CpfpInput.P2tr> {
  const parts = s.split(':')
  if (parts.length !== 4) {
    throw new Error(`Invalid funding UTXO '${s}', expected txid:vout:value:pubkey`)
  }
  const txid = parts[0]
  const vout = parseInt(parts[1], 10)
  if (isNaN(vout)) {
    throw new Error(`Invalid vout in '${s}'`)
  }
  const value = BigInt(parts[2])
  const pubkey = parts[3]

  if (kind === 'p2wpkh') {
    return new CpfpInput.P2wpkh({ txid, vout, value, pubkey })
  }
  return new CpfpInput.P2tr({ txid, vout, value, pubkey })
}

function formatExitTransactions(response: any): string[] {
  const lines: string[] = []
  lines.push(
    `Recoverable ${response.recoverableValueSat} sats, ` +
    `total fee ${response.totalFeeSat} sats ` +
    `(cpfp ${response.cpfpFeeSat}, fanout ${response.fanoutFeeSat}, sweep ${response.sweepFeeSat}), ` +
    `${response.transactions.length} transaction(s):`
  )

  for (let i = 0; i < response.transactions.length; i++) {
    const tx = response.transactions[i]
    const after = tx.dependsOn.length > 0
      ? `, after ${tx.dependsOn.join(',')}`
      : ''
    const csv = tx.csvTimelockBlocks != null
      ? `, csv ${tx.csvTimelockBlocks} blocks`
      : ''
    lines.push(`  [${i}] ${tx.kind} status=${tx.status} txid=${tx.txid}${after}${csv}`)

    if (tx.status.tag === ExitTransactionStatus_Tags.Confirmed) {
      const blockHeight = tx.status.inner?.blockHeight
      if (blockHeight != null) {
        lines.push(`      (confirmed in block ${blockHeight}, nothing to broadcast)`)
      } else {
        lines.push('      (already confirmed, nothing to broadcast)')
      }
      continue
    }
    if (tx.status.tag === ExitTransactionStatus_Tags.WaitingForDependencies) {
      lines.push('      (waiting on the transactions it depends on)')
    }
    if (tx.status.tag === ExitTransactionStatus_Tags.WaitingForTimelock) {
      const spendableAtHeight = tx.status.inner?.spendableAtHeight
      if (spendableAtHeight != null) {
        lines.push(`      (waiting for its timelock, until block ${spendableAtHeight})`)
      } else {
        lines.push('      (waiting for its timelock)')
      }
    }

    const pkg = tx.cpfpTxHex
      ? `${tx.txHex},${tx.cpfpTxHex}`
      : tx.txHex
    lines.push(`      Package: ${pkg}`)
  }

  return lines
}

async function handleUnilateralExit(sdk: BreezSdkInterface, args: string[]): Promise<string> {
  const feeRateStr = parseFlag(args, '--fee-rate')
  const destination = parseFlag(args, '--destination')

  if (!feeRateStr || !destination) {
    return 'Usage: advanced unilateral-exit --fee-rate <rate> --destination <addr> [--funding-kind p2wpkh|p2tr] [--leaf <id>,<id>,...] [--utxo txid:vout:value:pubkey ...] [--secret-key <hex>] [--output-file <path>]'
  }

  const feeRate = BigInt(feeRateStr)
  const fundingKindStr = parseFlag(args, '--funding-kind') ?? 'p2tr'
  const leafIds = parseMultiFlag(args, '--leaf')
  const outputFile = parseFlag(args, '--output-file')

  const fundingKind = fundingKindStr === 'p2wpkh'
    ? new CpfpFundingKind.P2wpkh()
    : new CpfpFundingKind.P2tr()

  const selection = leafIds && leafIds.length > 0
    ? new ExitLeafSelection.Specific({ leafIds })
    : new ExitLeafSelection.Auto()

  const prepared = await sdk.prepareUnilateralExit({
    feeRateSatPerVbyte: feeRate,
    fundingKind,
    destination,
    selection,
  })

  const lines: string[] = [formatValue(prepared)]

  if (prepared.leaves.length === 0) {
    lines.push('No leaves to exit.')
    return lines.join('\n')
  }

  const utxoArgs = parseRepeatedFlag(args, '--utxo')
  if (utxoArgs.length === 0) {
    lines.push('No funding provided; showing the quote only.')
    lines.push('Provide --utxo txid:vout:value:pubkey and --secret-key <hex> to sign.')
    return lines.join('\n')
  }

  const secretKey = parseFlag(args, '--secret-key')
  if (!secretKey) {
    return 'Error: --secret-key is required when --utxo is provided'
  }

  const fundingInputs = utxoArgs.map(u => parseCpfpInput(u, fundingKindStr))
  const signer = singleKeyCpfpSigner(hexToArrayBuffer(secretKey.trim()))

  const response = await sdk.unilateralExit(
    { prepared, fundingInputs },
    signer
  )

  lines.push(...formatExitTransactions(response))

  if (outputFile) {
    const outputPath = resolvePath(outputFile)
    await writeExitFile(outputPath, response)
    lines.push(`Wrote the exit to ${outputPath}`)
  }

  return lines.join('\n')
}

// --- check-unilateral-exit ---

async function handleCheckUnilateralExit(sdk: BreezSdkInterface, args: string[]): Promise<string> {
  const inputFile = parseFlag(args, '--input-file')
  if (!inputFile) {
    return 'Usage: advanced check-unilateral-exit --input-file <path> [--output-file <path>]'
  }
  const outputFile = parseFlag(args, '--output-file')

  const inputPath = resolvePath(inputFile)
  const exit = await readExitFile(inputPath)

  const checked = await sdk.checkUnilateralExit({ exit })

  const lines: string[] = []
  lines.push(`Verdict: ${formatValue(checked.verdict)}`)
  if (checked.verdict.tag === UnilateralExitVerdict_Tags.Redo) {
    lines.push('  (this exit cannot be finished, quote and build it again)')
  }
  lines.push(...formatExitTransactions(checked.exit))

  const outputPath = resolvePath(outputFile ?? inputFile)
  await writeExitFile(outputPath, checked.exit)
  lines.push(`Wrote the exit to ${outputPath}`)

  return lines.join('\n')
}

// --- export-unilateral-exit-state ---

async function handleExportUnilateralExitState(sdk: BreezSdkInterface, args: string[]): Promise<string> {
  const outputFile = parseFlag(args, '--output-file')
  if (!outputFile) {
    return 'Usage: advanced export-unilateral-exit-state --output-file <path>'
  }

  const exported = await sdk.exportUnilateralExitState()
  const filePath = resolvePath(outputFile)
  await RNFS.writeFile(filePath, exported.exitState, 'utf8')

  return `Wrote ${exported.exitState.length} bytes to ${filePath}`
}

// --- import-unilateral-exit-state ---

async function handleImportUnilateralExitState(sdk: BreezSdkInterface, args: string[]): Promise<string> {
  const inputFile = parseFlag(args, '--input-file')
  if (!inputFile) {
    return 'Usage: advanced import-unilateral-exit-state --input-file <path>'
  }

  const filePath = resolvePath(inputFile)
  const exitState = await RNFS.readFile(filePath, 'utf8')

  const imported = await sdk.importUnilateralExitState({ exitState })

  return (
    `Imported ${imported.importedLeaves} leaf(s), ` +
    `skipped ${imported.skippedForeignLeaves} leaf(s) from a different wallet ` +
    `and ${imported.skippedConflictingLeaves} that disagree with what this wallet holds, ` +
    `left out the exit data of ${imported.skippedChains} leaf(s)`
  )
}

File: crates/breez-sdk/cli/src/command/advanced.rs
use std::fs;
use std::path::{Path, PathBuf};

use breez_sdk_spark::signer::single_key_cpfp_signer;
use breez_sdk_spark::{
    BreezSdk, CheckUnilateralExitRequest, CpfpFundingKind, CpfpInput, ExitLeafSelection,
    ExitTransactionStatus, ImportUnilateralExitStateRequest, PrepareUnilateralExitRequest,
    UnilateralExitRequest, UnilateralExitResponse, UnilateralExitVerdict,
};
use clap::{Subcommand, ValueEnum};
use rustyline::{Editor, history::DefaultHistory};

use crate::command::{CliHelper, print_value};

#[derive(Clone, Copy, Debug, ValueEnum)]
#[clap(rename_all = "lower")]
pub enum FundingKindArg {
    P2wpkh,
    P2tr,
}

impl From<FundingKindArg> for CpfpFundingKind {
    fn from(kind: FundingKindArg) -> Self {
        match kind {
            FundingKindArg::P2wpkh => CpfpFundingKind::P2wpkh,
            FundingKindArg::P2tr => CpfpFundingKind::P2tr,
        }
    }
}

/// Expert-only commands that build raw transactions for you to broadcast
/// yourself. Misuse can strand or lose funds.
#[derive(Clone, Debug, Subcommand)]
pub enum AdvancedCommand {
    /// Build and sign a unilateral exit. Quotes it first (which leaves, fees, how
    /// much to fund), then prompts for the funding UTXOs and signing key.
    UnilateralExit {
        /// Target fee rate in sat/vByte.
        #[arg(long)]
        fee_rate: u64,
        /// Funding UTXO kind.
        #[arg(long, value_enum, default_value_t = FundingKindArg::P2tr)]
        funding_kind: FundingKindArg,
        /// Destination address for the swept funds.
        #[arg(long)]
        destination: String,
        /// Leaf id to exit (repeatable). Omit to auto-select every profitable leaf.
        #[arg(long = "leaf")]
        leaf_ids: Vec<String>,
        /// File to write the signed exit to, for `check-unilateral-exit` to read
        /// back once its transactions are on-chain.
        #[arg(long)]
        output_file: Option<PathBuf>,
    },
    /// Read a signed exit written by `unilateral-exit` back against the chain:
    /// which of its transactions confirmed, what is ready to broadcast now, and
    /// whether the exit still holds.
    CheckUnilateralExit {
        /// File the exit was written to.
        #[arg(long)]
        input_file: PathBuf,
        /// File to write the updated exit to. Defaults to `--input-file`.
        #[arg(long)]
        output_file: Option<PathBuf>,
    },
    /// Export the wallet's unilateral exit state to a file, for safekeeping
    /// outside the wallet's own storage.
    ExportUnilateralExitState {
        /// File to write the exit state to.
        #[arg(long)]
        output_file: PathBuf,
    },
    /// Import a unilateral exit state previously written by
    /// `export-unilateral-exit-state`, merging it into the wallet.
    ImportUnilateralExitState {
        /// File the exit state was exported to.
        #[arg(long)]
        input_file: PathBuf,
    },
}

pub async fn handle_command(
    rl: &mut Editor<CliHelper, DefaultHistory>,
    sdk: &BreezSdk,
    command: AdvancedCommand,
) -> Result<bool, anyhow::Error> {
    match command {
        AdvancedCommand::UnilateralExit {
            fee_rate,
            funding_kind,
            destination,
            leaf_ids,
            output_file,
        } => {
            let prepared = sdk
                .prepare_unilateral_exit(PrepareUnilateralExitRequest {
                    fee_rate_sat_per_vbyte: fee_rate,
                    funding_kind: funding_kind.into(),
                    destination,
                    selection: exit_leaf_selection(leaf_ids),
                })
                .await?;
            print_value(&prepared)?;
            if prepared.leaves.is_empty() {
                println!("No leaves to exit.");
                return Ok(true);
            }

            let utxo_line = rl.readline(
                "Funding UTXO(s) as txid:vout:value:pubkey (space-separated, blank to stop): ",
            )?;
            if utxo_line.trim().is_empty() {
                println!("No funding provided; showing the quote only.");
                return Ok(true);
            }
            let funding_inputs = utxo_line
                .split_whitespace()
                .map(|u| parse_cpfp_input(u, funding_kind))
                .collect::<Result<Vec<_>, _>>()?;

            let key_line = rl.readline("Hex secret key for the funding UTXO(s): ")?;
            let signer = single_key_cpfp_signer(hex::decode(key_line.trim())?)?;

            let response = sdk
                .unilateral_exit(
                    UnilateralExitRequest {
                        prepared,
                        funding_inputs,
                    },
                    signer,
                )
                .await?;
            print_exit_transactions(&response);
            if let Some(output_file) = output_file {
                write_exit(&output_file, &response)?;
            }
            Ok(true)
        }
        AdvancedCommand::CheckUnilateralExit {
            input_file,
            output_file,
        } => {
            let exit = read_exit(&input_file)?;
            let checked = sdk
                .check_unilateral_exit(CheckUnilateralExitRequest { exit })
                .await?;
            println!("Verdict: {:?}", checked.verdict);
            if let UnilateralExitVerdict::Redo { .. } = checked.verdict {
                println!("  (this exit cannot be finished, quote and build it again)");
            }
            print_exit_transactions(&checked.exit);
            write_exit(output_file.as_deref().unwrap_or(&input_file), &checked.exit)?;
            Ok(true)
        }
        AdvancedCommand::ExportUnilateralExitState { output_file } => {
            let exported = sdk.export_unilateral_exit_state().await?;
            fs::write(&output_file, &exported.exit_state)?;
            println!(
                "Wrote {} bytes to {}",
                exported.exit_state.len(),
                output_file.display(),
            );
            Ok(true)
        }
        AdvancedCommand::ImportUnilateralExitState { input_file } => {
            let exit_state = fs::read_to_string(&input_file)?;
            let imported = sdk
                .import_unilateral_exit_state(ImportUnilateralExitStateRequest { exit_state })
                .await?;
            println!(
                "Imported {} leaf(s), skipped {} leaf(s) from a different wallet \
                 and {} that disagree with what this wallet holds, \
                 left out the exit data of {} leaf(s)",
                imported.imported_leaves,
                imported.skipped_foreign_leaves,
                imported.skipped_conflicting_leaves,
                imported.skipped_chains,
            );
            Ok(true)
        }
    }
}

/// Reads an exit written by [`write_exit`].
fn read_exit(path: &Path) -> Result<UnilateralExitResponse, anyhow::Error> {
    Ok(serde_json::from_str(&fs::read_to_string(path)?)?)
}

/// Writes the exit as JSON, the form [`read_exit`] reads back.
fn write_exit(path: &Path, exit: &UnilateralExitResponse) -> Result<(), anyhow::Error> {
    fs::write(path, serde_json::to_string_pretty(exit)?)?;
    println!("Wrote the exit to {}", path.display());
    Ok(())
}

/// Auto when no leaves are named, otherwise exactly the given leaves.
fn exit_leaf_selection(leaf_ids: Vec<String>) -> ExitLeafSelection {
    if leaf_ids.is_empty() {
        ExitLeafSelection::Auto
    } else {
        ExitLeafSelection::Specific { leaf_ids }
    }
}

/// Parses a `txid:vout:value:pubkey` funding UTXO into a [`CpfpInput`] of the
/// given kind. `pubkey` is hex; for P2TR it is the internal (untweaked) key.
fn parse_cpfp_input(s: &str, kind: FundingKindArg) -> Result<CpfpInput, anyhow::Error> {
    let [txid, vout, value, pubkey] = s.split(':').collect::<Vec<_>>()[..] else {
        return Err(anyhow::anyhow!(
            "invalid funding UTXO '{s}', expected txid:vout:value:pubkey"
        ));
    };
    let txid = txid.to_string();
    let vout = vout.parse::<u32>()?;
    let value = value.parse::<u64>()?;
    let pubkey = pubkey.to_string();
    Ok(match kind {
        FundingKindArg::P2wpkh => CpfpInput::P2wpkh {
            txid,
            vout,
            value,
            pubkey,
        },
        FundingKindArg::P2tr => CpfpInput::P2tr {
            txid,
            vout,
            value,
            pubkey,
        },
    })
}

/// Prints each exit transaction with a copy-pasteable `Package:` line: the tx
/// hex, plus its signed CPFP child when present, comma-separated. That is the
/// form mempool.space accepts for a package broadcast. Confirmed steps need no
/// broadcast, so they show no package.
fn print_exit_transactions(response: &UnilateralExitResponse) {
    println!(
        "Recoverable {} sats, total fee {} sats (cpfp {}, fanout {}, sweep {}), \
         {} transaction(s):",
        response.recoverable_value_sat,
        response.total_fee_sat,
        response.cpfp_fee_sat,
        response.fanout_fee_sat,
        response.sweep_fee_sat,
        response.transactions.len(),
    );
    for (i, tx) in response.transactions.iter().enumerate() {
        let after = if tx.depends_on.is_empty() {
            String::new()
        } else {
            format!(", after {}", tx.depends_on.join(","))
        };
        let csv = tx
            .csv_timelock_blocks
            .map(|b| format!(", csv {b} blocks"))
            .unwrap_or_default();
        println!(
            "  [{i}] {:?} status={:?} txid={}{after}{csv}",
            tx.kind, tx.status, tx.txid,
        );
        match tx.status {
            ExitTransactionStatus::Confirmed { block_height } => {
                match block_height {
                    Some(height) => {
                        println!("      (confirmed in block {height}, nothing to broadcast)");
                    }
                    None => println!("      (already confirmed, nothing to broadcast)"),
                }
                continue;
            }
            ExitTransactionStatus::WaitingForDependencies => {
                println!("      (waiting on the transactions it depends on)");
            }
            ExitTransactionStatus::WaitingForTimelock {
                spendable_at_height,
            } => match spendable_at_height {
                Some(height) => println!("      (waiting for its timelock, until block {height})"),
                None => println!("      (waiting for its timelock)"),
            },
            ExitTransactionStatus::Ready | ExitTransactionStatus::Unverified => {}
        }
        let package = match &tx.cpfp_tx_hex {
            Some(cpfp) => format!("{},{}", tx.tx_hex, cpfp),
            None => tx.tx_hex.clone(),
        };
        println!("      Package: {package}");
    }
}

#[cfg(test)]
mod tests {
    use breez_sdk_spark::{UnilateralExitLeaf, UnilateralExitTransaction, UnilateralExitTxKind};

    use super::*;

    /// Covers every status variant, both funding kinds, and each optional field
    /// set and unset, so a field that stops round-tripping is caught here.
    fn sample_exit() -> UnilateralExitResponse {
        UnilateralExitResponse {
            recoverable_value_sat: 100_000,
            total_fee_sat: 1_500,
            cpfp_fee_sat: 900,
            fanout_fee_sat: 400,
            sweep_fee_sat: 200,
            leaves: vec![UnilateralExitLeaf {
                leaf_id: "leaf-1".to_string(),
                value: 100_000,
            }],
            transactions: vec![
                UnilateralExitTransaction {
                    kind: UnilateralExitTxKind::FanOut,
                    node_id: None,
                    txid: "aa".to_string(),
                    tx_hex: "0200".to_string(),
                    cpfp_tx_hex: None,
                    csv_timelock_blocks: None,
                    depends_on: vec![],
                    status: ExitTransactionStatus::Confirmed {
                        block_height: Some(101),
                    },
                },
                UnilateralExitTransaction {
                    kind: UnilateralExitTxKind::Node,
                    node_id: Some("node-1".to_string()),
                    txid: "bb".to_string(),
                    tx_hex: "0201".to_string(),
                    cpfp_tx_hex: Some("0301".to_string()),
                    csv_timelock_blocks: Some(144),
                    depends_on: vec!["aa".to_string()],
                    status: ExitTransactionStatus::Confirmed { block_height: None },
                },
                UnilateralExitTransaction {
                    kind: UnilateralExitTxKind::Refund,
                    node_id: Some("node-1".to_string()),
                    txid: "cc".to_string(),
                    tx_hex: "0202".to_string(),
                    cpfp_tx_hex: Some("0302".to_string()),
                    csv_timelock_blocks: Some(2016),
                    depends_on: vec!["bb".to_string()],
                    status: ExitTransactionStatus::WaitingForTimelock {
                        spendable_at_height: Some(245),
                    },
                },
                UnilateralExitTransaction {
                    kind: UnilateralExitTxKind::Refund,
                    node_id: Some("node-2".to_string()),
                    txid: "dd".to_string(),
                    tx_hex: "0203".to_string(),
                    cpfp_tx_hex: None,
                    csv_timelock_blocks: None,
                    depends_on: vec!["bb".to_string()],
                    status: ExitTransactionStatus::WaitingForDependencies,
                },
                UnilateralExitTransaction {
                    kind: UnilateralExitTxKind::Node,
                    node_id: Some("node-3".to_string()),
                    txid: "ee".to_string(),
                    tx_hex: "0204".to_string(),
                    cpfp_tx_hex: None,
                    csv_timelock_blocks: None,
                    depends_on: vec![],
                    status: ExitTransactionStatus::Unverified,
                },
                UnilateralExitTransaction {
                    kind: UnilateralExitTxKind::Sweep,
                    node_id: None,
                    txid: "ff".to_string(),
                    tx_hex: "0205".to_string(),
                    cpfp_tx_hex: None,
                    csv_timelock_blocks: None,
                    depends_on: vec!["cc".to_string(), "dd".to_string()],
                    status: ExitTransactionStatus::Ready,
                },
            ],
            funding_inputs: vec![
                CpfpInput::P2tr {
                    txid: "11".to_string(),
                    vout: 0,
                    value: 5_000,
                    pubkey: "02ab".to_string(),
                },
                CpfpInput::P2wpkh {
                    txid: "22".to_string(),
                    vout: 3,
                    value: 7_000,
                    pubkey: "03cd".to_string(),
                },
                CpfpInput::Custom {
                    txid: "33".to_string(),
                    vout: 1,
                    value: 9_000,
                    script_pubkey_hex: "5120ef".to_string(),
                    signed_input_weight: 230,
                },
            ],
        }
    }

    /// What `unilateral-exit` writes is what `check-unilateral-exit` reads.
    #[test]
    fn exit_file_round_trips() {
        let dir = tempfile::tempdir().expect("tempdir");
        let path = dir.path().join("exit.json");
        let exit = sample_exit();

        write_exit(&path, &exit).expect("write");
        let read_back = read_exit(&path).expect("read");

        // Debug rather than the serialized form: a field that serializes away
        // on both sides would compare equal as JSON while losing its value.
        assert_eq!(format!("{exit:?}"), format!("{read_back:?}"));
    }
}
